reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.reporter.reporters
+
+import com.here.ort.model.Identifier
+import com.here.ort.model.LicenseFindingsMap
+import com.here.ort.model.ScanResult
+import com.here.ort.model.processStatements
+import com.here.ort.model.removeGarbage
+import com.here.ort.reporter.ReporterInput
+import com.here.ort.spdx.LICENSE_FILENAMES
+import com.here.ort.utils.CopyrightStatementsProcessor
+import com.here.ort.utils.FileMatcher
+import com.here.ort.utils.getUserOrtDirectory
+import com.here.ort.utils.log
+import com.here.ort.utils.storage.FileArchiver
+import com.here.ort.utils.storage.LocalFileStorage
+import com.here.ort.utils.toHexString
+
+import java.io.File
+import java.nio.file.Path
+import java.security.MessageDigest
+
+/**
+ * Creates a notice file containing the licenses for all non-excluded projects and packages, listed by their identifier.
+ * Detected copyrights are listed next to the licenses.
+ *
+ * If a [FileArchiver] is configured for the scanner the reporter will try to download any archived license files and
+ * include them in the notice file. Licenses detected in those license files will not be listed again, instead all
+ * copyrights associated to those licenses will be listed below the license file's content.
+ */
+class NoticeByPackageReporter : AbstractNoticeReporter() {
+    override val reporterName = "NoticeByPackage"
+    override val defaultFilename = "NOTICE_BY_PACKAGE"
+
+    override fun createProcessor(input: ReporterInput): NoticeProcessor = NoticeByPackageProcessor(input)
+}
+
+class NoticeByPackageProcessor(input: ReporterInput) : AbstractNoticeReporter.NoticeProcessor(input) {
+    companion object {
+        val DEFAULT_ARCHIVE_DIR by lazy { getUserOrtDirectory().resolve("scanner/archive") }
+        const val LICENSE_SEPARATOR = "\n  --\n\n"
+    }
+
+    override fun process(data: AbstractNoticeReporter.NoticeReportData): List<() -> String> =
```

Comment:
Good point, I have changed a preceding commit to rename `NoticeReport` to `NoticeReportModel` instead of `NoticeReportData` and adapted all variable names.